### PR TITLE
Tutorial

### PR DESCRIPTION
**Pull request by Danwhy**
Adds tutorial, example and resources on Property Based Testing with Elixir.

ref https://github.com/dwyl/learn-elixir/issues/93

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,278 @@
...
+```
+
+Now that we have our properties defined, let's write some code to test.
+
+``` elixir
+defmodule Frame do
+  def build(input) do
+    longest = Enum.max_by(input, fn s -> String.length(s) end) |> String.length()
+
+    top = String.duplicate("*", longest + 4)
```
Comment (samhstn) [inline README.md:144]: I'm not sure `top` is a good name here (used on the bottom too), maybe `border`?

File: README.md [hunk 2]
```diff
@@ -1 +1,278 @@
...
+    ...
+  end
+end
+```
+
+Now that we have our properties defined, let's write some code to test.
+
+``` elixir
+defmodule Frame do
+  def build(input) do
```
Comment (samhstn) [inline README.md:141]: I think `strings` might be a better arg name than `input`

File: README.md [hunk 3]
```diff
@@ -1 +1,278 @@
...
+    longest = Enum.max_by(input, fn s -> String.length(s) end) |> String.length()
```
Comment (samhstn) [inline README.md:142]: Maybe update this to `Enum.max_by(input, &String.length/1)` as you've done in the code

File: README.md [hunk 4]
```diff
@@ -1 +1,278 @@
...
+
+But what happens if your function receives an _unexpected_ input?
+
+You can't write an example based test for an example you haven't thought of. That's where property based testing comes in. Instead of testing examples, you define and test the properties of your code.
+
+For example, if you were writing some code to validate a credit card number, one of the properties would be that it would only accept numbers as an input. Then, a property based test would validate that any other input returned an error.
+
+## What?
+
+Property Based Testing originated with [QuickCheck](https://en.wikipedia.org/wiki/QuickCheck), a Haskell library created in 1999. QuickCheck takes assertions about about logical properties that a program should fulfill, then generates test cases that try to fail these assertions.
```
Comment (SimonLab) [inline README.md]: > takes assertions about about 

What about one less?

File: README.md [hunk 5]
```diff
@@ -1 +1,278 @@
...
+    longest = Enum.max_by(input, fn s -> String.length(s) end) |> String.length()
+
+    top = String.duplicate("*", longest + 4)
+
+    input
+    |> Enum.map(fn s ->
+      "* " <> s <> suffix(longest - String.length(s) + 1)
+    end)
+    |> List.insert_at(0, top)
+    |> List.insert_at(-1, top)
```
Comment (samhstn) [inline README.md:151]: I think I would have used `&String.pad_trailing/2` for this

```elixir
[border]
|> Enum.concat(Enum.map(strings, fn s -> "* " <> String.pad_trailing(s, longest - 4) <> " *") end)
|> Enum.concat([border])
```

(and remove the `suffix` function)

File: README.md [hunk 6]
```diff
@@ -1 +1,278 @@
...
+
+- The outputted frame should have a height that is the length of the list of strings, plus two for the top and bottom line of asterisks.
+- The frame should have a width that is the length of the longest string, plus four for the two spaces and two asterisks on either side of the word.
+- Each line of the frame should be of equal width.
+
+There are a few other properties we could add for completeness, such as what should happen if a non-string is passed, but we'll stick with these logical ones for now.
+
+So now we translate those properties into code. As mentioned above, we'll be using the StreamData library to generate test cases.
+
+The first thing we need to so is set up our test suite. Create a new elixir project using `mix new frame`, then put the following in your `test/frame_test.exs` file:
```
Comment (SimonLab) [inline README.md]: > The first thing we need to so

```suggestion
The first thing we need to do is set up our test suite. Create a new elixir project using `mix new frame`, then put the following in your `test/frame_test.exs` file:
```

File: example/lib/frame.ex [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    end)
+    |> List.insert_at(0, top)
+    |> List.insert_at(-1, top)
+  end
+
+  defp validate(input) do
+    case Enum.all?(input, &is_ascii?/1) do
+      true -> input
+      false -> raise ArgumentError
+    end
```
Comment (samhstn) [inline example/lib/frame.ex:30]: `if else` feels nicer here (only my personal preference though)

**Review by samhstn — COMMENTED** [review]
Such a great tutorial Dan, definitely gonna be using property based testing in future!

Thanks for the great intro to it ❤️

(just added my personal syntax preferences)

File: README.md (hunk 4 above)
Comment (Danwhy) [reply README.md]:
```suggestion
Property Based Testing originated with [QuickCheck](https://en.wikipedia.org/wiki/QuickCheck), a Haskell library created in 1999. QuickCheck takes assertions about logical properties that a program should fulfill, then generates test cases that try to fail these assertions.
```

File: README.md (hunk 3 above)
Comment (Danwhy) [reply README.md:142]: I actually started with `&String.length/1` on both, but took it out here to make it easier for beginners to understand. I should have updated them both though

File: example/lib/frame.ex (hunk 7 above)
Comment (nelsonic) [reply example/lib/frame.ex:30]: I'd tend to agree, but not a "blocker" to merging this PR. 
(_feel free to PR the change, happy to accept changes/simplifications_)

File: README.md (hunk 2 above)
Comment (nelsonic) [reply README.md:141]: PR welcome. 😉

File: README.md (hunk 1 above)
Comment (nelsonic) [reply README.md:144]: Good suggestion. PR welcome. 👍

**Review by nelsonic — APPROVED** [review]
@Danwhy wow! 😍

**Pull request merged**